### PR TITLE
Sprockets 4 Compatibility

### PR DESCRIPTION
**Pull request by schneems**
I'm able to get a really simple test app working with this gem with these changes, but so far haven't gotten the tests to pass.

### CODE REVIEW

**Comment by bolandrm** [comment]
cool, i may be able to help with the tests if needed.  LMK

**Comment by bolandrm** [comment]
it looks like the issue might be with switching to `env.register_transformer`.   it looks like the failed tests are trying to use the regular sass instead of sassc.

be careful that your example app is using sassc - i usually double check by intentionally creating a syntax error and seeing that it is `SassC::SyntaxError` instead of `Sass::SyntaxError`

**Comment by schneems** [comment]
In my example app I added a syntax error and it looks like it's coming from `sassc`

```
$ rm -rf tmp/cache/; rm -rf public/assets/; time RAILS_ENV=production rake assets:precompile
trash: tmp/cache/: path does not exist
trash: public/assets/: path does not exist
rake aborted!
SassC::SyntaxError: Error: Invalid CSS after "...ry-color: #333;": expected 1 selector or at-rule, was "$$$$LKJASLDKJFOISDJ"
        on line 2 of app/assets/stylesheets/foo.scss
>> $primary-color: #333;
   ---------------------^
/Users/richardschneeman/Documents/projects/sassc-rails-app/app/assets/config/manifest.js:3
/Users/richardschneeman/.gem/ruby/2.4.0/gems/sassc-1.9.0/lib/sassc/engine.rb:43:in `render'
/Users/richardschneeman/Documents/projects/sassc-rails/lib/sassc/rails/template.rb:34:in `block in call'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/utils.rb:137:in `module_include'
/Users/richardschneeman/Documents/projects/sassc-rails/lib/sassc/rails/template.rb:33:in `call'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:84:in `call_processor'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:66:in `block in call_processors'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:65:in `reverse_each'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:65:in `call_processors'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:22:in `block in <class:CompositeProcessor>'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:33:in `call'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:84:in `call_processor'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:66:in `block in call_processors'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:65:in `reverse_each'
/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:65:in `call_processors'
```

That was a good debugging idea. I'll have to dig into these test failures a bit more.

**Comment by schneems** [comment]
I did a `bundle open` added some debugging and ran tests. 

It looks like __no__ transformer is running, not the sassc-rails one nor the built in Sprockets one on these 9 failing tests.

In the loader right after we set processors I added

```
puts "======"
puts unloaded.compressed_path.inspect
puts processors.inspect
```

When I run sprockets tests I get results like this

```
======
"test/fixtures/sass/paths.scss?type=text/css&pipeline=self"
[#<Proc:0x007f94233c1248@/Users/richardschneeman/Documents/projects/sprockets/test/sprockets_test.rb:84>, #<struct Sprockets::ProcessorUtils::CompositeProcessor processor_strategy=#<Proc:0x007f9422070b08@/Users/richardschneeman/Documents/projects/sprockets/lib/sprockets/processor_utils.rb:22 (lambda)>, param=[#<Sprockets::Preprocessors::DefaultSourceMap:0x007f9423459188>, #<Sprockets::DirectiveProcessor:0x007f9423482df8 @header_pattern=/\A(?:(?m:\s*)(?:(?:\/\/.*\n?)+|(?:\/\*(?m:.*?)\*\/)))+/>, Sprockets::ScssProcessor], processors=[#<Sprockets::Preprocessors::DefaultSourceMap:0x007f9423459188>, #<Sprockets::DirectiveProcessor:0x007f9423482df8 @header_pattern=/\A(?:(?m:\s*)(?:(?:\/\/.*\n?)+|(?:\/\*(?m:.*?)\*\/)))+/>, Sprockets::ScssProcessor]>, #<Sprockets::DirectiveProcessor:0x007f9422ac6eb8 @header_pattern=/\A(?:(?m:\s*)(?:(?:\/\/.*\n?)+|(?:\/\*(?m:.*?)\*\/)))+/>, Sprockets::FileReader]
```

You can see where the `Sprockets::ScssProcessor` is being used.

However when I do the same with the scss-rails test suite I get results like this:

```
======
"app/assets/stylesheets/syntax_error.scss?type=text/scss"
[#<Sprockets::DirectiveProcessor:0x007fa1dac3e410 @header_pattern=/\A(?:(?m:\s*)(?:(?:\/\/.*\n?)+|(?:\/\*(?m:.*?)\*\/)))+/>, Sprockets::FileReader]
```

I think the difference is that Sprockets 4 will actually render your `.scss` file. Not the difference in the `type=` between the two one is `text/css` which works correctly the other is `text/scss` which isn't triggering the transformers.

So

```
new_css_output = render_asset("glob_test.scss")
``` 

In the tests should be 

```
new_css_output = render_asset("glob_test.css")
```

**Comment by bolandrm** [comment]
Ah, that makes sense.  Is that extension change backwards compatible?

**Comment by schneems** [comment]
Got all tests green. Did have to use both `register_transformer` and `register_engine` to support Sprockets 3. Details of why are in the commit.

**Comment by schneems** [comment]
Anything else to do here? Any questions?

File: lib/sassc/rails/railtie.rb [hunk 1]
```diff
@@ -50,8 +50,15 @@ class Railtie < ::Rails::Railtie
           self.sass_config = app.config.sass
         end
 
-        env.register_engine '.sass', SassC::Rails::SassTemplate
-        env.register_engine '.scss', SassC::Rails::ScssTemplate
+        if env.respond_to?(:register_transformer)
+          env.register_transformer 'text/sass', 'text/css', SassC::Rails::SassTemplate.new #->() { puts "yoyoyoy" }
+          env.register_transformer 'text/scss', 'text/css', SassC::Rails::ScssTemplate.new #->() { puts "yoyoyoy" }
```
Comment (samjewell) [inline lib/sassc/rails/railtie.rb]: Did you mean to leave the `puts "yoyoyoy"` in at the end of these lines?

Comment (samjewell) [inline lib/sassc/rails/railtie.rb]: Yeah I'll have a go tonight (UK time) 👍

**Comment by jankeesvw** [comment]
I just added issue #93, I think it's related.

**Review by fly938 — COMMENTED** [review]
1

**Pull request merged**